crates/rust-analyzer/src/main_loop.rs

```diff
@@ -486,6 +488,27 @@ impl GlobalState {
             // Only retry requests that haven't been cancelled. Otherwise we do unnecessary work.
             Task::Retry(req) if !self.is_completed(&req) => self.on_request(req),
             Task::Retry(_) => (),
+            Task::FileIndexState(uri) => {
+                let _p = profile::span("run_unindexed_project");
+
+                tracing::debug!(changes = self.process_changes(), "processes changes");
+                let snap = self.snapshot();
+                let id =
+                    crate::lsp::from_proto::file_id(&snap, &uri).expect("Unable to get file ID");
+
+                if let Ok(crates) = &snap.analysis.crates_for(id) {
+                    if crates.is_empty() {
+                        tracing::debug!(?uri, "rust-analyzer does not track this file");
+                        self.send_notification::<ext::UnindexedProject>(
+                            ext::UnindexedProjectParams {
+                                text_documents: vec![lsp_types::TextDocumentIdentifier { uri }],
+                            },
+                        );
+                    } else {
+                        tracing::warn!("was unable to get analysis for crate")
+                    }
+                }
```

Comment:
Now we are still processing it on the main thread :D All of this should be done in the spawned task on the thread pool, where as `Task::FileIndexState` should be used to return the result of the computation here (the `snap.analysis.crates_for(id)` result) and then have the task handler here just send the notification.

Reply:
After chatting about this over DMs, I'm _not_ going to take the suggested approach and make this noop task more substantial since there is no way to ensure that this code runs _after_ `GlobalState::process_changes()` has been called without doing one of the following:
- Add a retry mechanism for tasks, which would make this a de-facto spin loop. This might be okay!
- Creating a mechanism of enqueuing a task, but deferring spawning them until _after_ a turn of `GlobalState::handle_event`. This ensures that `GlobalState::process_changes()` has been called.